lib/checkstl.cpp

```diff
@@ -3117,6 +3117,70 @@ void CheckStl::knownEmptyContainer()
     }
 }
 
+void CheckStl::eraseIteratorOutOfBoundsError(const Token *ftok, const Token* itertok, const ValueFlow::Value* val)
+{
+    const std::string func = ftok ? ftok->str() : std::string("erase");
+    const std::string iter = itertok ? itertok->expressionString() : std::string("it");
+
+    const bool isConditional = val && val->isPossible();
+    std::string msg;
+    if (isConditional) {
+        msg = ValueFlow::eitherTheConditionIsRedundant(val->condition) + " or function '" + func + "()' is called on the iterator '" + iter + "' which is out of bounds.";
+    } else {
+        msg = "Calling function '" + func + "()' on the iterator '" + iter + "' which is out of bounds.";
+    }
+
+    reportError(ftok, isConditional ? Severity::warning : Severity::error,
+                "eraseIteratorOutOfBounds",
+                msg, CWE628, Certainty::normal);
+}
+
+static const ValueFlow::Value* getIterValue(const Token* tok, ValueFlow::Value::ValueType vt)
+{
+    auto it = std::find_if(tok->values().begin(), tok->values().end(), [&](const ValueFlow::Value& v) {
+        return v.valueType == vt && (v.isPossible() || v.isKnown());
+    });
+    return it != tok->values().end() ? &*it : nullptr;
+}
+
+void CheckStl::eraseIteratorOutOfBounds()
+{
+    logChecker("CheckStl::eraseIteratorOutOfBounds");
+    for (const Scope *function : mTokenizer->getSymbolDatabase()->functionScopes) {
+        for (const Token *tok = function->bodyStart; tok != function->bodyEnd; tok = tok->next()) {
+
+            if (!tok->valueType())
+                continue;
+            const Library::Container* container = tok->valueType()->container;
+            if (!container)
+                continue;
+            if (!Token::Match(tok->next(), ". %name% ("))
```

Comment:
would it make any sense to handle arbitrary container expressions in this checker? my suggestion would be to check something like: `Token::Match(tok->astParent(), ". %name% (") && tok->astParent()->astOperand1() == tok`

Reply:
example code.. I don't know why there would be extra parentheses:

    std::vector<int> v;
    (v).erase(v.begin());

Reply:
It seems that these parentheses are removed already. I have found an example that didn't work, but that required changes to `getArgumentStart()` as well.

A quick search shows 250+ instances of `%name% (` in our code base...